cmd/algorand-indexer/daemon.go

```diff
@@ -65,276 +67,338 @@ var (
-					filepath.Join(indexerDataDir, autoLoadIndexerConfigName))
-				panic(exit{1})
-			}
-			// No config file supplied via command line, auto-load it
-			configs, err := os.Open(configFile)
-			if err != nil {
-				maybeFail(err, "%v", err)
-			}
-			defer configs.Close()
-			err = viper.ReadConfig(configs)
-			if err != nil {
-				maybeFail(err, "invalid config file (%s): %v", viper.ConfigFileUsed(), err)
-			}
+func loadIndexerParamConfig(cfg *daemonConfig) error {
+	var err error
+	// If someone supplied a configuration file but also said to enable all parameters,
+	// that's an error
+	if cfg.suppliedAPIConfigFile != "" && cfg.enableAllParameters {
+		err = errors.New("not allowed to supply an api config file and enable all parameters")
+		logger.WithError(err).Errorf("API Parameter Error: %v", err)
+		return err
+	}
+	autoloadParamConfigPath := filepath.Join(cfg.indexerDataDir, autoLoadParameterConfigName)
+	paramConfigFound := util.FileExists(autoloadParamConfigPath)
+	// If we auto-loaded configs but a user supplied them as well, we have an error
+	if paramConfigFound {
+		if cfg.suppliedAPIConfigFile != "" {
+			err = fmt.Errorf("api parameter configuration was found in data directory (%s) as well as supplied via command line.  Only provide one",
+				filepath.Join(cfg.indexerDataDir, autoLoadParameterConfigName))
+			logger.WithError(err).Errorf("indexer parameter config error: %v", err)
+			return err
 		}
+		cfg.suppliedAPIConfigFile = autoloadParamConfigPath
+		logger.Infof("Auto-loading parameter configuration file: %s", suppliedAPIConfigFile)
+	}
+	return err
+}
 
-		if paramConfigFound {
-			if suppliedAPIConfigFile != "" {
-				logger.Errorf(
-					"api parameter configuration was found in data directory (%s) as well as supplied via command line.  Only provide one.",
-					filepath.Join(indexerDataDir, autoLoadParameterConfigName))
-				panic(exit{1})
-			}
-			suppliedAPIConfigFile = filepath.Join(indexerDataDir, autoLoadParameterConfigName)
-			fmt.Printf("Auto-loading parameter configuration file: %s", suppliedAPIConfigFile)
+func createIndexerPidFile(cfg *daemonConfig) error {
+	var err error
+	logger.Infof("Creating PID file at: %s\n", cfg.pidFilePath)
+	fout, err := os.Create(cfg.pidFilePath)
+	if err != nil {
+		err = fmt.Errorf("%s: could not create pid file, %v", cfg.pidFilePath, err)
+		logger.WithError(err).Error(err)
+		return err
+	}
+	_, err = fmt.Fprintf(fout, "%d", os.Getpid())
+	if err != nil {
+		err = fmt.Errorf("%s: could not write pid file, %v", cfg.pidFilePath, err)
+		logger.WithError(err).Error(err)
+		return err
+	}
+	err = fout.Close()
+	if err != nil {
+		err = fmt.Errorf("%s: could not close pid file, %v", cfg.pidFilePath, err)
+		logger.WithError(err).Error(err)
+		return err
+	}
+	return err
+}
 
-		}
+func runDaemon(daemonConfig *daemonConfig) error {
+	var err error
+	config.BindFlagSet(daemonConfig.flags)
+	err = configureLogger()
+	if err != nil {
```

Comment:
You can avoid reusing / shadowing the err object by using this pattern more
```suggestion
	if err := configureLogger(); err != nil {
```